### ISSUE TITLE
Bluestore encryption, wal an db

### ISSUE THREAD

**Issue by swiftgist** [SUSE/DeepSea#1138]
### Description of Issue/Question
While encryption of Bluestore is supported, we have a nonsensical approach.  The ceph-disk utility refuses to use existing partitions for encryption support.  Instead, the wal and db size are found in ceph.conf.  Generally, this means all wal and db partitions will be the same size throughout the cluster since the method to do this specifically for any OSD in ceph.conf is painful to manage.

We allow the customization of the wal and db size for all of filestore and bluestore except with bluestore encryption.  The admin can only learn of this exception the hard way.  When the size of either is provided in the pillar, we log on the minion that we cannot support these values here but create encrypted Bluestore partitions with the default sized wal and db.  See #1136 

After the admin removes each of these OSDs, updates the ceph.conf (i.e. adds bluestore_block_db_size, bluestore_block_wal_size) and updates the pillar (i.e. removes the wal_size and db_size), then Stage 3 will do what the admin intended.

***
We should not create encrypted bluestore partitions if the wal or db size is specified in the pillar.

Stage 2 should warn the admin that encryption, db_size and wal_size will not work.  Bluestore encryption and db_size/wal_size are mutually exclusive.

The `Encrypted OSDs not supported` for is_incorrect should be check not implemented.  Otherwise, dozens of these messages confuse the admin.

The proposal runner always supplies and wal/db size.  This needs to be corrected for encryption.

**Comment by Martin-Weiss** [comment]
Can´t we take what is in the proposal for wal and db size and add this to each OSDs settings in the ceph.conf? 

I agree that the size of the ceph.conf might increase - but if deepsea manages this automatically it should not be too painfull for the admins..

(having the same size for WAL for all OSDs might be ok (2GB) - but for the DB I see that this really depends on the data-disk and DB disk sizes - so this should be possible to be adjusted per OSD..)

**Comment by Martin-Weiss** [comment]
> The ceph-disk utility refuses to use existing partitions for encryption support.

And an other option might be to enhance ceph-disk to accept pre-created partitions... not sure why this is not supported or to change deepsea to create all db/wal using ceph-disk instead of "pre-creating" them,

**Comment by swiftgist** [comment]
Now consider the management of what you just requested.  That ceph.conf would contain these settings for every OSD?  For a thousand or so OSDs?  Next, the addition of an OSD would indirectly cause all OSDs to restart?  (We cannot know the 1000+ options in ceph.conf and perfectly decide which options require restarts.  At the moment, we left the granularity at the component level.  To make one part an exception to this would be more development as well.)

To your second question, ceph-disk is going away.  I do not know if ceph-volume has the same issue.  From a configuration perspective, I believe having the partition sizes with the device name in a minion named file is generally easier to manage, understand and safer.  Having a single mammoth configuration file imposes many issues.  That's why we went this way for filestore and bluestore in general.  Encryption suddenly became an exception with a difficult to explain deployment.  We unfortunately did what we could at the time, but meeting a release date left this in the state it currently is.

Our current situation is still wrong though.  Although not ideal, the improvements above would at least eliminate a trip down a dead end path.

**Comment by Martin-Weiss** [comment]
> That ceph.conf would contain these settings for every OSD?

Yes  - as this is a one of setting - I do not see a problem as long as the management happens automated via salt. But I agree that the real solution is to use the central config store of ceph and not ceph.conf anymore - or even better - use the parameters given on the CLI for ceph-disk.
Customer requirement needs to be fulfilled anyway - and that is have different sizes for DB depending on the disk size / number of objects.. and we can not delegate "use the same DB size for all"...

>Next, the addition of an OSD would indirectly cause all OSDs to restart?

This would be a bug (maybe in the design?). Why should I restart OSD.1 if I change a setting for OSD.2?

In case ceph-disk is deprecated and we have to use ceph-volume - next step should be to see if that fixes the problem and allows different WAL/DB sizes per OSD even in case encryption is used.
